HIVE-25449: datediff() gives wrong output when run in a tez task with some non-UTC timezone

ShubhamChaurasia opened the pull request:
Please find the repro and detailed description here - https://issues.apache.org/jira/browse/HIVE-25449

### What changes were proposed in this pull request?
Removed usage of SimpleDateFormat which was using local timezone parsing scalar dates. 


### Why are the changes needed?
This fixes the bug described in the Jira and datediff() now gives correct values.

### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Manual testing and added two new unit tests - testDateDiffColScalarWithTz() and testDateDiffScalarColWithTz()

abstractdog commented:
LGTM, I would request only minor changes, otherwise, it's +1:
there are another UDFs using formatter, please check those also, especially VectorUDFDateDiffColCol which seems to do the same wrong behavior
```
      date.setTime(formatter.parse(string).getTime());
      output.vector[i] = DateWritableV2.dateToDays(date);
```

ShubhamChaurasia commented:
> LGTM, I would request only minor changes, otherwise, it's +1:
> there are another UDFs using formatter, please check those also, especially VectorUDFDateDiffColCol which seems to do the same wrong behavior
> 
> ```
>       date.setTime(formatter.parse(string).getTime());
>       output.vector[i] = DateWritableV2.dateToDays(date);
> ```
> 
> also if you want to use timezone hack in qtest, feel free to use this feature:
> https://github.com/apache/hive/blob/master/itests/util/src/main/java/org/apache/hadoop/hive/ql/qoption/QTestTimezoneHandler.java

Thanks @abstractdog for the review. Nice catch. I have corrected VectorUDFDateDiffColCol as well. Also added a qtest in addition to unit tests.

abstractdog commented:
+1
